### PR TITLE
Meaningful Windows Support

### PR DESCRIPTION
**Pull request by cr1901**
This PR is not yet ready. However, I would appreciate having the changes I've pushed so far approved before continuing.

## Caveats
Strictly speaking, with the changes I've currently pushed, `sby` now works on Windows. However, the support is limited to one task per job, because the poll loop relies on `select`. `select` does not work on `fd`s on Windows, so an alternate solution is required.

Additionally, I'm not able to test the `abc`, `btor` or `aiger` engines at present; in particular, `btor` was [more complex](https://github.com/YosysHQ/SymbiYosys/compare/master...cr1901:no-resource?expand=1#diff-79c4f1bed2788c3c9c3ff133517af7deR115) to add Windows support. Thus, I'm more concerned that my refactoring doesn't break the `btor` engine on Linux.

## Changes So Far
* On Windows, `fcntl` and `resource` modules are gated out. This means the steps setting each tasks' `stdout` to non-blocking doesn't run, and process running time information is always `00:00:00`. I know of workarounds for lack of `fcntl`; I'd rather just gate out process running time on Windows.
* When constructing task command lines, use the OS to determine which command separator to use. On POSIX OSes, use `;`. On Windows, use `&` (equivalent to `&&` in POSIX).

## Next Steps
To complete this PR, I have to figure out how to handle the lack of `select` on fds to implement polling. If running multiple tasks, the current Windows behavior right now is for tasks to run sequentially rather than interleaved (but all processes are spawned. So all but one task blocks waiting to send its output back to `sby`).

### CODE REVIEW

File: sbysrc/sby_engine_btor.py [hunk 1]
```diff
@@ -105,11 +105,19 @@ def exit_callback(retcode):
...
-                    setupcmd = "cd %s;" % (job.workdir)
                     finalwit = "engine_%d/trace.wit" % engine_idx
                 else:
-                    setupcmd = "cd %s; { echo sat; btorsim --states model/design_btor.btor engine_%d/trace.wit; } > engine_%d/simtrace.wit &&" % (job.workdir, engine_idx, engine_idx)
+                    if os.name == "posix":
+                        setupcmd = setupcmd + "{ echo sat; btorsim --states model/design_btor.btor engine_%d/trace.wit; } > engine_%d/simtrace.wit &&" %
+                            (engine_idx, engine_idx)
+                    else:
+                        # Windows has no concept of grouping commands into one stream.
+                        # Create and append to the final witness in sequence instead.
```
Comment (cliffordwolf) [inline sbysrc/sby_engine_btor.py]: According to https://en.wikibooks.org/wiki/Windows_Batch_Scripting#Redirection its `( ... ) > file` on windows.

**Comment by cliffordwolf** [comment]
> On Windows, use & (equivalent to && in POSIX).

According to https://en.wikibooks.org/wiki/Windows_Batch_Scripting#Syntax `&` on win is equivalent to `;` on POSIX and `&&` has the same meaning on both architectures.

It looks like most of the patch could be replaced by simply `y/{};/()&/` on all commands when on windows.  Maybe that could be done just once in `SbyTask`, drastically simplifying the patch.

**Comment by cr1901** [comment]
>It looks like most of the patch could be replaced by simply y/{};/()&/ on all commands when on windows. Maybe that could be done just once in SbyTask, drastically simplifying the patch.

I agree- commands before being passed to SbyTask are "normalized" to `sh`, and are converted to the appropriate OS format during `__init__`. I'll update and force push.

## Select Replacement Proposal
Re: the `select` loop, I have a few options, and would like feedback if you have a preference:

### TCP/IP Sockets
Each time a `SbyTask` is started on Windows, spawn a thread from SymbiYosys to do _blocking_ `readline` calls from the subprocess' `stdout` and redirect to a TCP/IP socket (using ephemeral ports- we get about 16k of them; if you run out of those, you have bigger problems :)). 

`select` _does_ work on TCP/IP sockets on Windows, so run `select` on sockets instead of fds. This might not even require `if os.name == "posix"` gating! Line [102](https://github.com/YosysHQ/SymbiYosys/blob/master/sbysrc/sby_core.py#L102) of `sby_core.py` is replaced by a nonblocking read on each socket each time data comes through. Each blocking thread and socket dies when the subprocess exits.

### Explicit Event Queue
Replace the [select call](https://github.com/YosysHQ/SymbiYosys/blob/master/sbysrc/sby_core.py#L191) by a blocking read to an event queue (`queue.Queue`) that lasts `1.0` seconds before timing out. As with the socket approach, each subprocess spawns a blocking thread to wait for each subprocess' `stdout`. 

However, each thread then writes each line of  `stdout` to a per-thread queue, _and then_ puts a dummy event into the event queue to cause the main [taskloop](https://github.com/YosysHQ/SymbiYosys/blob/master/sbysrc/sby_core.py#L180) to keep going.
[102](https://github.com/YosysHQ/SymbiYosys/blob/master/sbysrc/sby_core.py#L102) of `sby_core.py` becomes a nonblocking read to each per-thread queue.

The disadvantage here is that while AFAICT, Python `Queue`s have atomic updates, the combination of putting data in each per-thread `stdout` queue and then updating the event queue is non atomic. It's possible for the event queue to be empty while there is actually data to read from each per-thread queue when control returns to the main SymbiYosys thread. 

I don't think the above is a huge problem at a glance. The main consequence seems to be that spurious iterations of the task loop that accomplish no work will happen occasionally. This _already_ happens when `select` times out in the current code. I would appreciate corrections if I'm missing something.

### The Hard Way
Bypass Python and call the Windows API directly from Python using `ctypes` and the actual `HANDLE` from `fileno`. `WaitForMultipleObjects` is the `select` equivalent. `NamedPipes` is analogous to converting a blocking file descriptor to non-blocking using `fcntl`. Even though this has all the nonblocking functionality of the *nix implementation of `taskloop`, I _really_ don't want to do this :).

**Comment by cliffordwolf** [comment]
> TCP/IP Sockets

No. That's a horrible solution.

> Explicit Event Queue | Replace the select call by a blocking read ...

No. I'd rather not support windows at all than use a hack like this as main mechanism to launch and monitor processes.

> The Hard Way | I really don't want to do this :).

Then don't. ;) Afaiu it works fine in WSL and I don't think anyone else actually wants to to run sby in plain windows, and it's not something that we would planning to ever "officially support" anyway.

**Comment by whitequark** [comment]
I would say that using WSL is an awful hack, and not supporting Windows is a portability failure.

What's the problem with using asyncio on Linux and Windows to monitor subprocesses? If you use ProactorEventLoop, asyncio does all the `WaitForMultipleObjects` stuff for you transparently, as well as dealing with pipes.

**Comment by cr1901** [comment]
Yes, using `asyncio` would also solve these problems.

`asyncio` is part of Python 3.5 and later, officially. 3.4 includes it on a provisional basis, and 3.3 can use it via a [package](https://pypi.org/project/asyncio/). I think 3.3 is a reasonable minimum supported version (and 3.3 minimum doesn't make obsolete either of the recent PRs I've made to `YosysHQ` to make Windows support more ergonomic).

**Comment by cr1901** [comment]
@cliffordwolf Just FYI aside from the async situation, I'm done all other platform-specific things I wanted to add. In short, I force-pushed and put all OS-string replacement in `SbyTask` to greatly simplify the patch/ Then I added:

* A comment :).
* A message for the process time being unavailable on Windows.
* A property to the JUnit schema that writes the value of `os.name`.

**Pull request merged**